File: examples/e2e-consent-demo/agent/consentAwareCall.ts
// The agent's consent-aware tool call — the whole of Part D's behaviour in one
// function.
//
// The agent itself has no consent logic: it never decides what the user may
// authorize, it only notices when an SP says "not without a grant" and hands
// off to that SP's own consent page. What makes step 5 of the demo work is the
// first line of doWork(): before building any presentation, ask the wallet
// whether a grant for THIS (service, user) pair already exists. If it does, it
// travels in the VP and the SP is satisfied on the first attempt — no prompt,
// no second grant issued.

import { VPBuilder, type SignedVC, type SignedVP } from '@helixid/core';
import type { AgentWallet } from '@helixid/sdk-js';

export interface ConsentPrompt {
  serviceDid: string;
  consentUrl: string;
  requiredScope: string;
}

/**
 * Surfaces consent to the End User and returns the grant the SP issued, or
 * null if the user declined. In the browser demo this opens the SP's consent
 * page; in tests it drives the same two HTTP routes directly.
 */
export type ConsentHandler = (prompt: ConsentPrompt) => Promise<SignedVC | null>;

export interface CallSpToolOptions {
  wallet: AgentWallet;
  /** DID or email — must be the identifier the grant captured at consent time. */
  userDid: string;
  spMcpUrl: string;
  serviceDid: string;
  toolName: string;
  args?: Record<string, unknown>;
  onConsentRequired: ConsentHandler;
  /**
   * Stitches every audit event this call produces — presentation,
   * verification, authorization, invocation, and any grant issued along the
   * way — into one traceable chain for a single user action.
   */
  correlationId?: string;
}

export interface CallSpToolResult {
  ok: boolean;
  data?: Record<string, unknown>;
  /** True when this call had to stop and ask the user to authorize. */
  consentPrompted: boolean;
  error?: { code: string; reason?: string; message: string };
}

export class ConsentDeclinedError extends Error {
  constructor(serviceDid: string) {
    super(`User declined consent for ${serviceDid}`);
    this.name = 'ConsentDeclinedError';
  }
}

interface JsonRpcResponse {
  result?: { structuredContent?: Record<string, unknown> };
  error?: {
    code?: number;
    message?: string;
    data?: { code?: string; reason?: string; requiredScope?: string; consentUrl?: string };
  };
}

/** The wallet's existing standing grant for this SP and this user, if any. */
function findExistingGrant(
  wallet: AgentWallet,
  serviceDid: string,
  userDid: string,
): SignedVC | undefined {
  const stored = wallet.selectGrant(serviceDid, userDid);
  // selectGrant returns the wallet's metadata row, not a parsed credential.
  return stored ? (JSON.parse(stored.vcJson) as SignedVC) : undefined;
}

async function buildVP(
  wallet: AgentWallet,
  serviceDid: string,
  userDid: string,
  grant: SignedVC | undefined,
): Promise<SignedVP> {
  const agentVC = wallet.credentials.find((vc) =>
    (vc.type as string[]).includes('HelixAgentCredential'),
  );
  if (!agentVC) {
    throw new Error('Agent wallet holds no HelixAgentCredential. Run enrollment first.');
  }

  return new VPBuilder({
    // Grant is always the second, independent entry — never merged into the
    // agent VC's delegation chain.
    credentials: grant ? [agentVC, grant] : [agentVC],
    holderDid: wallet.getDID(),
    targetService: serviceDid,
    userDid,
  }).sign(wallet.getPrivateKeyHex(), `${wallet.getDID()}#key-1`);
}

async function postToolCall(
  spMcpUrl: string,
  toolName: string,
  args: Record<string, unknown>,
  vp: SignedVP,
  correlationId?: string,
): Promise<JsonRpcResponse> {
  const response = await fetch(spMcpUrl, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({
      jsonrpc: '2.0',
      id: 1,
      method: 'tools/call',
      params: {
        name: toolName,
        arguments: {
          ...args,
          _helixVP: vp,
          ...(correlationId !== undefined ? { _helixCorrelationId: correlationId } : {}),
        },
      },
    }),
  });
  return (await response.json()) as JsonRpcResponse;
}

export async function callSpTool(options: CallSpToolOptions): Promise<CallSpToolResult> {
  const { wallet, userDid, spMcpUrl, serviceDid, toolName } = options;
  const args = options.args ?? {};

  // Step 5 hinges on this: reuse a standing grant if the wallet already has one.
  const existingGrant = findExistingGrant(wallet, serviceDid, userDid);
  const first = await postToolCall(
    spMcpUrl,
    toolName,
    args,
    await buildVP(wallet, serviceDid, userDid, existingGrant),
    options.correlationId,
  );

  if (!first.error) {
    return { ok: true, consentPrompted: false, ...(first.result?.structuredContent !== undefined ? { data: first.result.structuredContent } : {}) };
  }

  if (first.error.data?.code !== 'CONSENT_REQUIRED') {
    return {
      ok: false,
      consentPrompted: false,
      error: {
        code: first.error.data?.code ?? 'CALL_FAILED',
        ...(first.error.data?.reason !== undefined ? { reason: first.error.data.reason } : {}),
        message: first.error.message ?? 'Tool call failed',
      },
    };
  }

  // The SP wants a grant. Hand off to its consent page.
  const grantVC = await options.onConsentRequired({
    serviceDid,
    consentUrl: first.error.data.consentUrl ?? '',
    requiredScope: first.error.data.requiredScope ?? '',
  });
  if (!grantVC) {
    throw new ConsentDeclinedError(serviceDid);
  }

  await wallet.addCredential(grantVC);

  const retry = await postToolCall(
    spMcpUrl,
    toolName,
    args,
    await buildVP(wallet, serviceDid, userDid, grantVC),
    options.correlationId,
  );

  if (retry.error) {
    return {
      ok: false,
      consentPrompted: true,
      error: {
        code: retry.error.data?.code ?? 'CALL_FAILED',
        ...(retry.error.data?.reason !== undefined ? { reason: retry.error.data.reason } : {}),
        message: retry.error.message ?? 'Tool call failed after consent',
      },
    };
  }

  return {
    ok: true,
    consentPrompted: true,
    ...(retry.result?.structuredContent !== undefined ? { data: retry.result.structuredContent } : {}),
  };
}

File: examples/e2e-consent-demo-py/agent/consent_aware_call.py
# The agent's consent-aware tool call -- the whole of Part D's behaviour in
# one function. Python port of agent/consentAwareCall.ts.
#
# The agent itself has no consent logic: it never decides what the user may
# authorize, it only notices when an SP says "not without a grant" and hands
# off to that SP's own consent page. What makes step 5 of the demo work is
# the first line of call_sp_tool(): before building any presentation, ask the
# wallet whether a grant for THIS (service, user) pair already exists.

from __future__ import annotations

import json
from typing import Any, Callable, Dict, Optional

import requests

from helix_sdk.vp_builder import VPBuilder
from helix_sdk.wallet import AgentWallet


class ConsentDeclinedError(Exception):
    def __init__(self, service_did: str) -> None:
        super().__init__(f"User declined consent for {service_did}")
        self.name = "ConsentDeclinedError"


def _find_existing_grant(wallet: AgentWallet, service_did: str, user_did: str) -> Optional[Dict[str, Any]]:
    stored = wallet.select_grant(service_did, user_did)
    return json.loads(stored.vc_json) if stored else None


def _build_vp(wallet: AgentWallet, service_did: str, user_did: str, grant: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    agent_vc = next((vc for vc in wallet.credentials if "HelixAgentCredential" in (vc.get("type") or [])), None)
    if not agent_vc:
        raise RuntimeError("Agent wallet holds no HelixAgentCredential. Run enrollment first.")

    credentials = [agent_vc, grant] if grant else [agent_vc]
    builder = VPBuilder(
        credentials=credentials, holder_did=wallet.get_did(), target_service=service_did, user_did=user_did
    )
    return builder.sign(wallet.get_private_key_hex(), f"{wallet.get_did()}#key-1")


def _post_tool_call(
    sp_mcp_url: str, tool_name: str, args: Dict[str, Any], vp: Dict[str, Any], correlation_id: Optional[str] = None
) -> Dict[str, Any]:
    arguments = {**args, "_helixVP": vp}
    if correlation_id is not None:
        arguments["_helixCorrelationId"] = correlation_id
    response = requests.post(
        sp_mcp_url,
        json={"jsonrpc": "2.0", "id": 1, "method": "tools/call", "params": {"name": tool_name, "arguments": arguments}},
        timeout=15,
    )
    return response.json()


# Called with {serviceDid, consentUrl, requiredScope}; returns the grant VC
# or None if the user declined. In the browser demo this always returns
# None -- the server never blocks; the browser drives the SP consent popup
# and retries POST /api/call itself.
ConsentHandler = Callable[[Dict[str, str]], Optional[Dict[str, Any]]]


def call_sp_tool(
    wallet: AgentWallet,
    user_did: str,
    sp_mcp_url: str,
    service_did: str,
    tool_name: str,
    on_consent_required: ConsentHandler,
    args: Optional[Dict[str, Any]] = None,
    correlation_id: Optional[str] = None,
) -> Dict[str, Any]:
    args = args or {}

    # Step 5 hinges on this: reuse a standing grant if the wallet already has one.
    existing_grant = _find_existing_grant(wallet, service_did, user_did)
    first = _post_tool_call(
        sp_mcp_url, tool_name, args, _build_vp(wallet, service_did, user_did, existing_grant), correlation_id
    )

    if not first.get("error"):
        result = {"ok": True, "consentPrompted": False}
        content = (first.get("result") or {}).get("structuredContent")
        if content is not None:
            result["data"] = content
        return result

    error = first["error"]
    error_data = error.get("data") or {}
    if error_data.get("code") != "CONSENT_REQUIRED":
        return {
            "ok": False,
            "consentPrompted": False,
            "error": {
                "code": error_data.get("code", "CALL_FAILED"),
                "reason": error_data.get("reason"),
                "message": error.get("message", "Tool call failed"),
            },
        }

    # The SP wants a grant. Hand off to its consent page.
    grant_vc = on_consent_required(
        {
            "serviceDid": service_did,
            "consentUrl": error_data.get("consentUrl", ""),
            "requiredScope": error_data.get("requiredScope", ""),
        }
    )
    if not grant_vc:
        raise ConsentDeclinedError(service_did)

    wallet.add_credential(grant_vc)

    retry = _post_tool_call(
        sp_mcp_url, tool_name, args, _build_vp(wallet, service_did, user_did, grant_vc), correlation_id
    )

    if retry.get("error"):
        retry_error = retry["error"]
        retry_data = retry_error.get("data") or {}
        return {
            "ok": False,
            "consentPrompted": True,
            "error": {
                "code": retry_data.get("code", "CALL_FAILED"),
                "reason": retry_data.get("reason"),
                "message": retry_error.get("message", "Tool call failed after consent"),
            },
        }

    result = {"ok": True, "consentPrompted": True}
    content = (retry.get("result") or {}).get("structuredContent")
    if content is not None:
        result["data"] = content
    return result
